### PR TITLE
build: update workflow to support ts

### PR DESCRIPTION
**Pull request by trim21**
This PR doesn't migrate source to TS except `errors.ts` as example ( or tester ), full migration and refacting can be done in seprated PRs.

It migrate source files to ESM and update tool config to support typescript.

close https://github.com/minio/minio-js/issues/834 https://github.com/minio/minio-js/issues/835

# Code Style ( excluding already described in #1120 )

Add a set of eslint rules to force import style to make it work with build script.

`import` should be fully specified to import source files, developers of this project should write `import {} from './errors.ts'` for `errors.ts`, `import {} from './helpers.js'` for `helpers.js`, file extension must be included, this is essential for testing and building

function comment:

should always write jsdoc/tsdoc, tsc will only emit jsdoc/tsdoc in type definition. function/method comment `// ...` won't work.

# Building

to not break [existing](https://github.com/ultimaup/ultima/blob/1264a109c6ee87fc2dbdd1566969772dd70b46c8/file-manager/minio.js#L5-L6) `require('minio/dist/main/...')`, we still need to do a file to file transpiling.

There is no existing bundler can do this ( and tsc doesn't fit this case ), so a script `build.mjs` is added to use babel to transform codes.

`src/` directory will be inclued in npm package but user should not be able to import then. editor (at lease vscode and WebStorm) will be able to jump to implementation (raw ts files in `src/` dir) with tsconfig `"declarationMap": true`

Type definition are also imported from `@types/minio`, need some extra work to keep it up to date.

## ESM

Also build esm mjs files to `./dist/esm/`

Use babel to transpile type definition from `.d.ts` to `.d.mts`, so importing esm files will also be typed.

Add named export for all default export, and suggest stop make any new default export (keep existing backward compatible and remove them in next major version), use named exports only, like what we did in the package entrypoint. This is more clear to support both cjs and esm.

current building already transform esm default export to cjs default property, so no behavoir change here:
```js
// current npm version
const r = require('minio/dist/main/CredentialProvider.js');

console.log(r.default);
```

now users can write code in a better way:
```js
import { CredentialProvider } from 'minio/dist/esm/CredentialProvider.mjs';

const { CredentialProvider } = require('minio/dist/main/CredentialProvider.js');
console.log(new CredentialProvider(...));

// this still work, keep backward compatibility
const cp = require('minio/dist/main/CredentialProvider.js');
console.log(new cp.default(...));
```

# Testing

still use mocha, no test cases are changed, but add `@babel/register` to load esm and ts files, so no need to compile before testing.

`.env` file is totally optional

## mint

mint support https://github.com/minio/mint/pull/356

# About `error.ts`

`Error.captureStackTrace` should be used when you create a plain object `{}` as `Error` and want to add stack, not to be used when you extending built-in `Error`.

There is no need to call `Error.captureStackTrace` in the constructor of Error or sub-class of Erro, only need to set `error.name`.

Also it doesn't support es2022 standard [`error.cause`](https://github.com/tc39/proposal-error-cause) (available in node16+, and doesn't broken lower version), there is no need to use it anymore.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -48,28 +67,39 @@
...
+    "prettier": "^2.8.7",
+    "rollup-plugin-node-externals": "^5.1.2",
     "source-map-support": "^0.5.13",
     "split-file": "^2.2.2",
-    "superagent": "^5.1.0",
-    "uuid": "^3.3.2"
+    "superagent": "^8.0.9",
+    "ts-node": "^10.9.1",
+    "typescript": "^5.0.4",
+    "uuid": "^9.0.0"
```
Comment (aldy505) [inline package.json:102]: we should really remove uuid later on. use `randomUUID` from `node:crypto` instead

**Review by aldy505 — COMMENTED** [review]
LGTM

File: package.json (hunk 1 above)
Comment (aldy505) [reply package.json:102]: ah, we're doing CI tests on node 12. 

it's okay then.

**Comment by prakashsvmx** [comment]
This is a great first step. Thank you for the contribution.  :+1: 

Some comments: 
The tests should be runnable locally  and externally (mint link below)
hope we need to add `.env` or `.env.local`   Please add and commit these files as well may be empty so others can follow it.


Code formatting can be taken up as a last step once we have all the other ts files migration. 
We could add and generate the types  once migration is complete, may be using `https://github.com/Swatinem/rollup-plugin-dts`  


I still need to do the test/validate (changes if any) with https://github.com/minio/mint/tree/master/run/core/minio-js

**Comment by prakashsvmx** [comment]
Another observation is : while running tests locally, the tests are not progressing after some point
e.g 
```
....
 listIncompleteUploads removeIncompleteUpload
      ✔ initiateNewMultipartUpload(bucketName, objectName, metaData, cb)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, objectName:datafile-65-MB, metaData:[object Object]
      - listIncompleteUploads(bucketName, prefix, recursive)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, prefix:datafile-65-MB, recursive: true_
      - listIncompleteUploads(bucketName, prefix, recursive)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, recursive: true_
      - removeIncompleteUploads(bucketName, prefix)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, prefix:datafile-65-MB_
    fPutObject fGetObject
      ✔ fPutObject(bucketName, objectName, filePath, callback)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB_ (449ms)
      ✔ fPutObject(bucketName, objectName, filePath, metaData, callback)_bucketName:minio-js-test-d20bfea7-a775-460e-afa1-c1a04224e4cb, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB, metaData: [object Object]_ (323ms)

```
Steps to replicate:
add `.env` file with 
```
SERVER_ENDPOINT="localhost:22000" 
ACCESS_KEY="minio" 
SECRET_KEY="minio123"
```
```
# Run Server in a terminal
CI=true  MINIO_ROOT_USER=minio MINIO_ROOT_PASSWORD=minio123 minio server /tmp/log{1...4}  --address ":22000" --console-address ":9012"

# Run build
rm -rf node_modules  dist && npm install && npm run build

# Run tests in another terminal
npm run test

```

**Comment by trim21** [comment]
can we add Code formatting first？would be easy for contributors to unify code formatting and pr diff could look better

**Comment by trim21** [comment]
> The tests should be runnable locally and externally (mint link below)

I'll take a look at mint later

> hope we need to add .env or .env.local Please add and commit these files as well may be empty so others can follow it.

I add a `.env.example` file, it would be better to not commit `.env` file into git so user can change it and have a different config

**Comment by trim21** [comment]
> Another observation is : while running tests locally, the tests are not progressing after some point

I find these tests become unexpected slow, I'll look into it

```text
    fPutObject fGetObject
      √ fPutObject(bucketName, objectName, filePath, callback)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB, filePath:C:\Users\Trim21\AppData\Local\Temp/datafile-65-MB_ (88069ms)
      √ fPutObject(bucketName, objectName, filePath, metaData, callback)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB, filePath:C:\Users\Trim21\AppData\Local\Temp/datafile-65-MB, metaData: [object Object]_ (55277ms)
      √ fGetObject(bucketName, objectName, filePath, callback)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB, filePath:C:\Users\Trim21\AppData\Local\Temp/datafile-65-MB.download_ (19084ms)
      √ removeObject(bucketName, objectName, filePath, callback)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB_ (184ms)
      √ fPutObject(bucketName, objectName, filePath, metaData)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB, filePath:C:\Users\Trim21\AppData\Local\Temp/datafile-65-MB_ (62562ms)
      √ fGetObject(bucketName, objectName, filePath)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB, filePath:C:\Users\Trim21\AppData\Local\Temp/datafile-65-MB.download_ (16623ms)
      √ removeObject(bucketName, objectName, filePath, callback)_bucketName:minio-js-test-1ef393ec-7c3a-41d2-873e-323764721bd3, objectName:datafile-65-MB_ (191ms)
```

**Comment by prakashsvmx** [comment]
@trim21  the master the tests are running  fine.

<details>
```
.....
    Force Deletion of prefix
      Test force removal of a prefix
        ✓ putObject(bucketName, objectName, stream)_bucketName:minio-js-fd-nv-e940a55c-9db6-43cf-9744-03d927484735, objectName:my-prefix/datafile-100-kB, stream:100Kib_
        ✓ removeObject(bucketName, objectList, removeOpts)_bucketName:minio-js-fd-nv-e940a55c-9db6-43cf-9744-03d927484735_Remove 1 object
        ✓ listObjects(bucketName, prefix, recursive)_bucketName:minio-js-fd-nv-e940a55c-9db6-43cf-9744-03d927484735, prefix: 'my-prefix', recursive:true_


  409 passing (34s)
  3 pending
```
</details>

**Comment by trim21** [comment]
should be fixed I think

```text

  407 passing (44s)
  3 pending

```

**Comment by aldy505** [comment]
I wanted to suggest the usage of minio:cicd-edge docker image as a service on the CI, rather than running the minio server as a background process. 

But the cicd-edge tag is far outdated and there's not much maintainers that wants to use it.

**Comment by trim21** [comment]
> I wanted to suggest the usage of minio:cicd-edge docker image as a service on the CI, rather than running the minio server as a background process.
> 
> But the cicd-edge tag is far outdated and there's not much maintainers that wants to use it.

I think this is not very related to this PR and can be a seprated issue ?

**Comment by aldy505** [comment]
>> I wanted to suggest the usage of minio:cicd-edge docker image as a service on the CI, rather than running the minio server as a background process.
>>
>> But the cicd-edge tag is far outdated and there's not much maintainers that wants to use it.
>
> I think this is not very related to this PR and can be a seprated issue ?

Yeah, but it's something that I'd suggest on the Typescript rewrite journey. I'll turn my issue on Typescript into a checklist later on.

**Comment by trim21** [comment]
I almost finish migrating all files except the main Client class. Just wait style pr and this pr to be merged, then I can submit them separated for easier code review

**Comment by trim21** [comment]
There is one more problem:

There are currently some project using minio@7 require files from `minio/dist/...`

https://github.com/ultimaup/ultima/blob/1264a109c6ee87fc2dbdd1566969772dd70b46c8/file-manager/minio.js#L5-L6

So to not make any breaking change, we still need to build them in cjs format instead of using any kind of bundler

https://github.com/minio/minio-js/pull/1119/commits/20768b60274a29ad6a2d105994b69989d1d422e3

**Comment by trim21** [comment]
This PR is ready for review now

**Comment by trim21** [comment]
I'm not sure how the release bot will build this project, So I add `npm run build` to `prepublishOnly` script to make sure package will build

**Comment by trim21** [comment]
mint support pull request created https://github.com/minio/mint/pull/356

**Comment by prakashsvmx** [comment]
> mint support pull request created [minio/mint#356](https://github.com/minio/mint/pull/356)

Thank you @trim21 . i was working in  parallel  on this. Sorry for not notifying you earlier.

**Comment by prakashsvmx** [comment]
Accidentally closed this.

**Comment by trim21** [comment]
> Accidentally closed this.

That's scaring, I thought you are working on this and decide to discard this PR. 😰

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**